pandas/core/arrays/string_arrow.py

```diff
@@ -227,10 +227,22 @@ def _chk_pyarrow_available(cls) -> None:
 
     @classmethod
     def _from_sequence(cls, scalars, dtype: Dtype | None = None, copy: bool = False):
+        from pandas.core.arrays.masked import BaseMaskedArray
+
         cls._chk_pyarrow_available()
-        # convert non-na-likes to str, and nan-likes to ArrowStringDtype.na_value
-        scalars = lib.ensure_string_array(scalars, copy=False)
-        return cls(pa.array(scalars, type=pa.string(), from_pandas=True))
+
+        if isinstance(scalars, BaseMaskedArray):
+            # avoid costly conversion to object dtype in ensure_string_array and
+            # numerical issues with Float32Dtype
+            na_values = scalars._mask
+            result = scalars._data
+            result = lib.ensure_string_array(result, copy=False, convert_na_value=False)
+            result[na_values] = ArrowStringDtype.na_value
```

Comment:
Alternatively, might also be able to do `pa.array(result, mask=mask, type=pa.string())`, then pyarrow will use the mask for missing values, and we don't need the setitem operation.

Reply:
done.